src/utils.rs

```diff
@@ -82,84 +82,73 @@ pub fn find_message_type_in_file_descriptor(
+/// Very similar to find_message_type_by_name, but it narrows down a list of file descriptors to the ones where package
+/// name matches the given package name, either the one provided explicitly, or the one from the file_descriptor.
 pub fn find_message_descriptor(
-  message_name: &str,
-  package: Option<&str>,
-  file_descriptor: &FileDescriptorProto,
-  all_descriptors: &HashMap<String, &FileDescriptorProto>
+    message_name: &str,
+    package: Option<&str>,
+    file_descriptor: &FileDescriptorProto,
+    all_descriptors: &HashMap<String, &FileDescriptorProto>,
 ) -> anyhow::Result<DescriptorProto> {
-  if let Some(package) = package {
-    find_message_with_package_in_file_descriptors(message_name, package, file_descriptor, all_descriptors)
-  } else {
-    find_message_type_in_file_descriptors(message_name, file_descriptor, all_descriptors)
-  }
+    let package = package
+        .filter(|p| !p.is_empty())
+        .unwrap_or(file_descriptor.package.as_deref().unwrap_or_default());
+    debug!(
+        "Looking for message '{}' in package '{}'",
+        message_name, package
+    );
+    find_all_file_descriptors_for_package(package, all_descriptors)
+        .unwrap()
+        .iter()
+        .find_map(|fd| find_message_type_in_file_descriptor(message_name, fd).ok())
+        .ok_or_else(|| {
+            anyhow!(
+                "Did not find a message type '{}' in any of the file descriptors for package '{}'",
+                message_name,
+                package,
+            )
+        })
 }
 
-/// Search for a message by type name and package in the file descriptor, and if not found,
-/// search in all the descriptors. This will first check with the package name, and if nothing is
-/// found, will then fall back to just using the message name
-pub fn find_message_with_package_in_file_descriptors(
-  message_name: &str,
-  package: &str,
-  file_descriptor: &FileDescriptorProto,
-  all_descriptors: &HashMap<String, &FileDescriptorProto>
-) -> anyhow::Result<DescriptorProto> {
-  find_file_descriptor_for_package(package, file_descriptor, all_descriptors)
-    .and_then(|file_descriptor| find_message_type_in_file_descriptor(message_name, &file_descriptor))
-    .or_else(|_| find_message_type_in_file_descriptors(message_name, file_descriptor, all_descriptors))
-}
-
-fn find_file_descriptor_for_package(
-  package: &str,
-  file_descriptor: &FileDescriptorProto,
-  all_descriptors: &HashMap<String, &FileDescriptorProto>
-) -> anyhow::Result<FileDescriptorProto> {
-  let package = if package.starts_with('.') {
-    &package[1..]
-  } else {
-    package
-  };
-  let found = if let Some(descriptor_package) = &file_descriptor.package {
-    if descriptor_package == package {
-      Some(file_descriptor)
+fn find_all_file_descriptors_for_package<'a>(
+    package: &str,
+    all_descriptors: &'a HashMap<String, &FileDescriptorProto>,
+) -> anyhow::Result<Vec<&'a FileDescriptorProto>> {
```

Comment:
Not sure about this return value. I tend to try avoid using lifetimes, unless I want to return a reference for performance reasons.

You are calling `cloned` on the iterator below, so the values are being cloned anyway, and then return a vector of references. Might as well as remove the references and lifetime and just clone the collection at the end.

Reply:
I'm getting a bit confused by this if I'm honest. Can you please provide an example? Rust ownership logic is very new to me.

Reply:
If I leave the code as is, and remove lifetime specified (which AI suggested lol), I'm getting this error:
```
missing lifetime specifier
this function's return type contains a borrowed value, but the signature does not say whether it is borrowed from `package` or one of `all_descriptors`'s 2 lifetimes
...
utils.rs(141, 45): consider introducing a named lifetime parameter: `<'a>`, `'a `, `'a `, `'a `, `'a `
```

Reply:
I'll try to explain.

The return type `Vec<&'a FileDescriptorProto>` is a vector of pointers to FileDescriptorProto objects. Rust forces you to define a lifetime for these pointers to guarantee that they will never live longer that then data they are pointing to (i.e. have dangling pointers). It does this by tying the lieftime of the pointers to the passed in HashMap `&'a HashMap<String, &FileDescriptorProto>`. So once the destructor for `all_descriptors` is run, the compiler will not let you use those pointers.

However, lower done the function `cloned()` is being called on the iteration, which clones every item in iteration (actually looking at the types in RustRover, it is just cloning the pointers). So if you're going to clone the items, you might as well just create a new Vector of cloned objects and return that. Then you don't need any lifetime specifiers.

So the function definition becomes:
```rust
fn find_all_file_descriptors_for_package(
    package: &str,
    all_descriptors: &HashMap<String, &FileDescriptorProto>,
) -> anyhow::Result<Vec<FileDescriptorProto>>
```
see, no pointers and no lifetimes

and lower down we change the `cloned()` call to something like

```rust
        .map(|fd| *fd.clone()) // fd here is a pointer to a pointer here, so we must first de-reference it
        .collect(); // No longer need the type defined here, Rust will infer it
```

Reply:
The reason that `cloned()` won't work, as the HashMap values are references, and the `values()` function returns an iterator over references to the values, which ends up being double references. You can see these if you use RustRover

![image](https://github.com/pactflow/pact-protobuf-plugin/assets/38309842/a0492ea9-87e3-45f8-ab2a-05f2fe19602f)

Reply:
thanks for the detailed explanation!

Reply:
This makes a lot of sense, if we're cloning the values we don't need lifetime specifiers, and if we're keeping lifetime, we should only be cloning references (which is what was happening?). 
However, when I change my code like so:
```
fn filter_file_descriptors<F>(all_descriptors: &HashMap<String, &FileDescriptorProto>, filter: F
) -> Vec<FileDescriptorProto>
where
    F: FnMut(&&&FileDescriptorProto) -> bool,
{
    all_descriptors.values()
        .filter(filter)
        .map(|d| *d.clone())
        .collect()
}
```
I'm getting the following error:
```
cannot move out of a shared reference
move occurs because value has type `FileDescriptorProto`, which does not implement the `Copy` trait
```
I'm sure I'm missing something here
<img width="719" alt="Screenshot 2024-06-05 at 6 40 24 PM" src="https://github.com/pactflow/pact-protobuf-plugin/assets/49661990/67ec4caa-3b53-4e1c-b3f4-0bafa005b62c">

Reply:
This compiles for me:
```
fn find_all_file_descriptors_with_no_package(
  all_descriptors: &HashMap<String, &FileDescriptorProto>
) -> anyhow::Result<Vec<FileDescriptorProto>> {
  let found = filter_file_descriptors(all_descriptors, |d| d.package.is_none());
  if found.is_empty() {
      Err(anyhow!("Did not find any file descriptors with no package specified"))
  } else {
      debug!("Found {} file descriptors with no package", found.len());
      Ok(found)
  }
}

fn filter_file_descriptors<F>(
    all_descriptors: &HashMap<String, &FileDescriptorProto>,
    filter: F
) -> Vec<FileDescriptorProto>
where
    F: FnMut(&&FileDescriptorProto) -> bool
{
    all_descriptors.values()
        .map(|fd| *fd) // Convert &&FileDescriptorProto -> &FileDescriptorProto
        .filter(filter)
        .cloned()
        .collect()
}
```

Reply:
Huh that's some high level pointer magic right there :) Thanks!